ruby/ql/test/library-tests/dataflow/array-flow/array_flow.rb

```diff
@@ -0,0 +1,523 @@
+    end
+    sink(b) # $ hasValueFlow=46.1
+end
+
+def m47
+    a = [0, 1, 2, source(47.1)]
+    a.fill(source(47.2), 1, 1)
+    sink(a[3]) # $ hasValueFlow=47.1 $ hasValueFlow=47.2
+    a.fill(source(47.3))
+    sink(a[0]) # $ hasValueFlow=47.3
+    a.fill do |i|
+        source(47.4)
+    end
+    sink(a[0]) # $ hasValueFlow=47.4
+    a.fill(2) do |i|
+        source(47.5)
+    end
+    sink(a[0]) # $ hasValueFlow=47.4 $ hasValueFlow=47.5
+end
+
+def m48
+    a = [0, 1, 2, source(48)]
+    b = a.filter do |x|
+        sink(x) # $ hasValueFlow=48
+    end
+    sink(b[0]) # $ hasValueFlow=48
+end
+
+def m49
+    a = [0, 1, 2, source(49)]
+    b = a.filter_map do |x|
+        sink(x) # $ hasValueFlow=49
+    end
+    sink(b[0]) # $ hasValueFlow=49
+end
+
+def m50
+    a = [0, 1, 2, source(50)]
+    b = a.filter! do |x|
+        sink(x) # $ hasValueFlow=50
+    end
+    sink(b[0]) # $ hasValueFlow=50
+end
+
+def m51
+    a = [0, 1, 2, source(51.1)]
+    b = a.find(-> { source(51.2) }) do |x|
+        sink(x) # $ hasValueFlow=51.1
+    end
+    sink(b) # $ hasValueFlow=51.1 $ hasValueFlow=51.2
+end
+
+def m52
+    a = [0, 1, 2, source(52)]
+    b = a.find_all do |x|
+        sink(x) # $ hasValueFlow=52
+    end
+    sink(b[0]) # $ hasValueFlow=52
+end
+
+def m53
+    a = [0, 1, 2, source(53)]
+    a.find_index do |x|
+        sink(x) # $ hasValueFlow=53
+    end
+end
+
+def m54(i)
+    a = [source(54.1), 1, 2, source(54.2)]
+    a[i] = source(54.3)
+    sink(a.first) # $ hasValueFlow=54.1 $ hasValueFlow=54.3
+    b = a.first(2)
+    sink(b[0]) # $ hasValueFlow=54.1 $ hasValueFlow=54.3
+    sink(b[4]) # $ hasValueFlow=54.3
+    c = a.first(i)
+    sink(c[0]) # $ hasValueFlow=54.1 $ hasValueFlow=54.3
+    sink(c[3]) # $ hasValueFlow=54.2 $ hasValueFlow=54.3
+end
+
+def m55
+    a = [0, 1, 2, source(55.1)]
+    b = a.flat_map do |x|
+        sink(x) # $ hasValueFlow=55.1
+        [x, source(55.2)]
+    end
+    sink(b[0]) # $ hasValueFlow=55.1 $ hasValueFlow=55.2
+end
+
+def m56
+    a = [0, 1, [2, source(56)]]
+    b = a.flatten
+    sink(b[0]) # $ hasValueFlow=56
+end
+
+def m57
+    a = [0, 1, [2, source(57)]]
+    sink(a[2][1]) # $ hasValueFlow=57
+    a.flatten!
+    sink(a[0]) # $ hasValueFlow=57
+    sink(a[2][1]) # $ SPURIOUS: hasValueFlow=57
+end
+
+def m58
+    a = [0, 1, 2, source(58.1)]
+    b = a.grep(/.*/)
+    sink(b[0]) # $ hasValueFlow=58.1
+    b = a.grep(/.*/) do |x|
+        sink x # $ hasValueFlow=58.1
+        source(58.2)
+    end
+    sink(b[0]) # $ hasValueFlow=58.2
+end
+
+def m59
+    a = [0, 1, 2, source(59.1)]
+    b = a.grep_v(/A/)
+    sink(b[0]) # $ hasValueFlow=59.1
+    b = a.grep_v(/A/) do |x|
+        sink x # $ hasValueFlow=59.1
+        source(59.2)
+    end
+    sink(b[0]) # $ hasValueFlow=59.2
+end
+
+def m60
+    a = [0, 1, 2, source(60)]
+    a.index do |x|
+        sink x # $ hasValueFlow=60
+    end
+end
+
+def m61
+    a = [0, 1, 2, source(61.1)]
+    a.initialize_copy([source(61.2)])
```

Comment:
I think the Ruby docs are a bit broken - `Array#initialize_copy` is a private method (at least on Ruby 3.0.2). There is `Array#replace` though.